app/models/openscap_result.rb

```diff
@@ -0,0 +1,68 @@
+require 'openscap'
+require 'openscap/ds/arf'
+require 'openscap/xccdf/benchmark'
+
+class OpenscapResult < ApplicationRecord
+  belongs_to :container_image
+  has_one    :binary_blob,           :dependent => :destroy, :autosave => true, :as => :resource
+  has_many   :openscap_rule_results, :dependent => :destroy, :autosave => true
+
+  OpenSCAP.oscap_init
+
+  def html
+    return unless binary_blob && binary_blob.binary
+    with_oscap_file(binary_blob.binary) do |arf_path|
+      arf = OpenSCAP::DS::Arf.new(arf_path)
+      arf.html
+    end
+  end
+
+  def raw=(value)
+    self.binary_blob = BinaryBlob.new(:name => 'openscap_compliance_arf', :data_type => 'XML')
+    binary_blob.binary = value
+    with_oscap_file(value) do |arf_path|
+      create_results(arf_path)
+    end
+  end
+
+  private
+
+  def create_results(arf_path)
+    results, meta = parse_arf(arf_path)
+
+    results.each do |k, v|
+      openscap_rule_results << OpenscapRuleResult.new(
+        :name            => ascii8bit_to_utf8(k),
+        :result          => ascii8bit_to_utf8(v.result),
+        :openscap_result => self,
+        :severity        => ascii8bit_to_utf8(meta[k].severity)
+      )
+    end
+  end
+
+  def parse_arf(arf_path)
+    arf = OpenSCAP::DS::Arf.new(arf_path)
+    rs = arf.test_result.rr
+    sds = arf.report_request
+    bench_source = sds.select_checklist!
+    [rs, OpenSCAP::Xccdf::Benchmark.new(bench_source).items]
```

Comment:
Right. So perhaps, we can divide this into methods. If we come up with useful names for the methods, things will clear up.

There is a similar code for this in Sat 6.
https://github.com/OpenSCAP/scaptimony/blob/master/app/models/scaptimony/arf_report_raw.rb#L22

I have put blob handling into a separate model. Does it make sense to do something similar. i.e. have a mixin to BinaryBlob, that provides public `to_html` and can provide `benchmark_items` (yours meta) and `rule_results`?